### PR TITLE
Schema for Users

### PR DESCRIPTION
**Pull request by anvial**
This PR adds User support for controller DB (Dqlite). The User abstraction is represented as several DB tables: 
- user
- user_authentication
- user_password
- user_activiation_key

This PR DOES NOT touch aspects of user authorization (state/useraccess.go) and all aspects of users in model DB, which will be provided as a separate commit. 

Note: `user_uuid` is used as a PRIMARY key.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
go test github.com/juju/juju/domain/schema/... -gocheck.v
```

## Links

**Jira card:** [JUJU-4894](https://warthogs.atlassian.net/browse/JUJU-4894)

[JUJU-4894]: https://warthogs.atlassian.net/browse/JUJU-4894?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: domain/schema/controller.go [hunk 1]
```diff
@@ -532,3 +533,46 @@ INSERT INTO autocert_cache_encoding VALUES
...
+    CONSTRAINT      fk_user_last_login_user
+        FOREIGN KEY (user_uuid)
+    REFERENCES      user(uuid)
+);
+
+CREATE TABLE user_password (
+    user_uuid       TEXT PRIMARY KEY,
+    password_hash   TEXT NOT NULL,
+    password_salt   TEXT NOT NULL,
+    disabled        BOOLEAN NOT NULL,
```
Comment (manadart) [inline domain/schema/controller.go:565]: Shouldn't this be on the `user` table?

File: domain/schema/controller.go [hunk 2]
```diff
@@ -532,3 +533,46 @@ INSERT INTO autocert_cache_encoding VALUES
...
+    created_by_uuid TEXT,
+    created_at      TIMESTAMP NOT NULL,
+    CONSTRAINT      fk_user_created_by_user
+        FOREIGN KEY (created_by_uuid)
+    REFERENCES      user(uuid)
+);
+
+CREATE UNIQUE INDEX idx_singleton_active_user ON user (name) WHERE deleted IS FALSE;
+
+CREATE TABLE user_last_login (
```
Comment (manadart) [inline domain/schema/controller.go:553]: This appears to have been a separate collection in Mongo in order to live outside the transaction operation workflow.

We don't need to mirror for Dqlite; just add `last_login` to the user table.

**Review by manadart — COMMENTED** [review]
Mostly OK; just a couple of changes.

File: domain/schema/controller.go (hunk 1 above)
Comment (anvial) [reply domain/schema/controller.go]: moved to the user table.

File: domain/schema/controller.go (hunk 2 above)
Comment (anvial) [reply domain/schema/controller.go]: Moved to the user table.

File: domain/schema/controller.go (hunk 1 above)
Comment (tlm) [reply domain/schema/controller.go]: I don't think disable should be on the user table. You are never disabling a user as that can only be done by removing the presence of a user.

What we are disabling is the users ability to authenticate with the api server. I think by that we should be keeping disable with the authentication information and not with the user.

File: domain/schema/controller.go (hunk 2 above)
Comment (tlm) [reply domain/schema/controller.go]: With the current ddl as we have it we have what I think is a cool concept in that we can have the notion of a user/owner in Juju for the purpose of ownership or user interfaces with the user necessarily having to authentication information.

My example would be let's say that you have a `jujud` snap where you can bootstrap a controller without a client we still need a owner of the model and everything else but not necessarily something that can or is logged in.

In my opinion the design we should be going for in the DDL is identity, authn, authz. We conflate identity and authn together by moving last_login onto the same table.

Comment (anvial) [reply domain/schema/controller.go]: Thx, know I hope the issue is resolved.

**Review by manadart — APPROVED** [review]
Thanks for the changes. Looks good now.

**Pull request merged**